rest/rest-client/src/main/java/org/ow2/proactive/scheduler/rest/SchedulerClient.java

```diff
@@ -803,26 +805,53 @@ public JobId submitFromCatalog(String catalogRestURL, String bucketName, String
+            Map<String, String> variables, Map<String, String> genericInfo)
             throws NotConnectedException, PermissionException, SubmissionClosedException, JobCreationException {
 
-        JobIdData jobIdData = null;
-
         HttpGet httpGet = new HttpGet(catalogRestURL + "/buckets/" + bucketName + "/resources/" + workflowName +
                                       "/raw");
-        httpGet.addHeader("sessionid", sid);
+        return submitFromCatalog(httpGet, variables, genericInfo);
 
-        try (CloseableHttpClient httpclient = HttpClients.createDefault();
-                CloseableHttpResponse response = httpclient.execute(httpGet)) {
+    }
 
-            jobIdData = restApiClient().submitXml(sid, response.getEntity().getContent(), variables, genericInfos);
-        } catch (Exception e) {
-            throwNCEOrPEOrSCEOrJCE(e);
-        }
-        return jobId(jobIdData);
+    @Override
+    public JobId submitFromCatalog(String catalogRestURL, String calledWorkflow)
+            throws NotConnectedException, PermissionException, SubmissionClosedException, JobCreationException {
+
+        return submitFromCatalog(catalogRestURL, calledWorkflow, new HashMap<>());
 
     }
 
+    @Override
+    public JobId submitFromCatalog(String catalogRestURL, String calledWorkflow, Map<String, String> variables)
+            throws NotConnectedException, PermissionException, SubmissionClosedException, JobCreationException {
+
+        return submitFromCatalog(catalogRestURL,
+                                 getBucketFromCalledWorkflow(calledWorkflow),
+                                 getNameFromCalledWorkflow(calledWorkflow),
+                                 variables,
+                                 null);
+    }
+
+    @Override
+    public JobId submitFromCatalog(String catalogRestURL, String calledWorkflow, Map<String, String> variables,
+            Map<String, String> genericInfo)
+            throws NotConnectedException, PermissionException, SubmissionClosedException, JobCreationException {
+
+        Optional<String> revision = getRevisionFromCalledWorkflow(calledWorkflow);
+
+        HttpGet httpGet = revision.map(r -> new HttpGet(catalogRestURL + "/buckets/" +
+                                                        getBucketFromCalledWorkflow(calledWorkflow) + "/resources/" +
+                                                        getNameFromCalledWorkflow(calledWorkflow) + "/revisions/" + r +
+                                                        "/raw"))
+                                  .orElseGet(() -> new HttpGet(catalogRestURL + "/buckets/" +
+                                                               getBucketFromCalledWorkflow(calledWorkflow) +
+                                                               "/resources/" +
+                                                               getNameFromCalledWorkflow(calledWorkflow) + "/raw"));
```

Comment:
Duplicate code can be removed, like this:

```
        Optional<String> revision = getRevisionFromCalledWorkflow(calledWorkflow);
        String workflowURL = catalogRestURL + "/buckets/" + getBucketFromCalledWorkflow(calledWorkflow) + "/resources/" + getNameFromCalledWorkflow(calledWorkflow);
        if (revision.isPresent()) {
            workflowURL +=  "/revisions/" + revision.get();
        }
        HttpGet httpGet = new HttpGet(workflowURL + "/raw");
```

But actually the 3 methods `getRevision`, `getBucket` and `getName` don't do much. I would replace them by a single one generating `workflowURL` directly.

Reply:
Thank you for noticing that duplication. I improved the code and removed the duplication.
However the methods `getBucketFromCalledWorkflow` and `getNameFromCalledWorkflow`  are used in two places. I would have to duplicate the code if remove them so I prefer to keep them.